subiquity/server/controllers/oem.py

```diff
@@ -72,14 +113,34 @@ async def load_metapackages_list(self, context) -> None:
                 except CommandNotFoundError:
                     self.model.metapkgs = []
                 else:
-                    self.model.metapkgs = await self.ubuntu_drivers.list_oem(
+                    metapkgs: List[str] = await self.ubuntu_drivers.list_oem(
                         root_dir=d.mountpoint,
                         context=context)
+                    self.model.metapkgs = [
+                        OEMMetaPkg(
+                            name=name,
+                            wants_oem_kernel=await self.wants_oem_kernel(
+                                name, context=context, overlay=d),
+                        ) for name in metapkgs]
+
         except OverlayCleanupError:
             log.exception("Failed to cleanup overlay. Continuing anyway.")
+
+        for pkg in self.model.metapkgs:
+            if pkg.wants_oem_kernel:
+                kernel_model = self.app.base_model.kernel
+                kernel_model.metapkg_name_override = flavor_to_pkgname(
+                        pkg.name, dry_run=self.app.opts.dry_run)
+
+                log.debug("overriding kernel flavor because of OEM")
```

Comment:
So the theory seems to be that if we're on OEM hardware, that we use the OEM kernel no matter what?  Even if configured otherwise by autoinstall or another method?  That doesn't sound correct.

It seems that the kernel detection logic should be overriding `linux-generic`, not any value of metapkg.

Reply:
> So the theory seems to be that if we're on OEM hardware, that we use the OEM kernel no matter what? 

If we're on OEM hardware, we will use the OEM flavor or the HWE flavor (depending on the OEM meta package). There's no way around it with the current design I'm afraid. 

> Even if configured otherwise by autoinstall or another method? That doesn't sound correct.

Correct, unless we disable that behavior in autoinstall with `oem: install: false`

Reply:
Then I think we need to raise a runtime error in that case.  The only way to satisfy the request to install a specific kernel is to disable the OEM functionality.  We can detect that scenario I think - if kernel supplied by autoinstall and we have an oem kernel metapackage name, raise and point out the fix (remove the kernel autoinstall directive or use `oem: install: false`).

Reply:
Ack that sounds fine to me.

Reply:
Implemented. That said, the error message still allows the install to continue because:
* oem controller cannot be part of the `install` set. It needs the apt overlay to operate properly
* oem controller is not a `postinstall` set. Doing so would not prevent the `install` from running.

We'd need an intermediate set that is required after the apt overlays are ready.

If that's okay I'll implement this in a subsequent PR